File: GoodsTracker/Communication/CommunicationUnit.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace GoodsTracker
{
    public enum ResultExec
    {
        EXEC_UNSUCCESS  = -3,
        INVALID_CMD     = -2,
        INVALID_PARAM   = -1,
        EXEC_SUCCESS    = 0,
    };

    internal delegate ResultExec onAnswerCmd(AnsCmd ans);

    abstract class CommunicationUnit
    {
        /*
         * Endereco da unidade
         */
        protected int   address;
        internal int    Address { get => address; set => address = value; }

        protected abstract void onReceiveAnswer(AnsCmd ans);

        internal CommunicationUnit(int val)
        {
            address = val;
            Communication.AddUnit(this);
        }

        internal Cmd createCMD(int dest, Operation o, string resource)
        {
            Cmd c               = new Cmd(resource, o);
            c.Header.Dest       = dest;
            c.Header.Address    = Master.ADDRESS;

            return c;
        }

        internal void sendCMD(Cmd cmd)
        {
            Communication.addCmd(cmd);
        }

        internal void processAnswer(AnsCmd ans)
        {
            try
            {
                // Pega o comando respectivo
                Cmd cmd = Communication.searchCmdOfAnswer(ans);

                if (cmd != null)
                {
                    // Executa evento de recebmento de resposta de comando
                    onReceiveAnswer(ans);

                    // Executa call back respectiva do comando
                    cmd.EventAnswerCmd?.Invoke(ans);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("Erro na execucao das callbacks de communicacao");
                Console.WriteLine(e.ToString());
                Debug.WriteLine(e.ToString());
            }
        }
    }
}

File: Mobile/GoodsTracker/app/src/main/java/com/example/jefferson/goodstracker/Communication/CommunicationUnit.java
package com.example.jefferson.goodstracker.Communication;

import android.util.Log;

import java.io.IOException;

/**
 * Created by Jefferson on 08/07/2017.
 */

abstract public class CommunicationUnit  extends Object implements ObserverCommunication {

    private static Communic communic = null;

    /*
     * Endereco da unidade
     */
    protected int   address;

    public CommunicationUnit(int val) throws IOException {

        address = val;

        Communication.getInstance().registerObserver(this);
    }

    public Cmd createCMD(String resource,Operation o){

        Cmd cmd  = new Cmd( CONST_COM.MASTER.ADDRESS,
                            address,
                            resource,
                            o);

        return cmd;
    }

    public void sendCMD(Cmd cmd) {

        Communication.getInstance().publish(cmd);
    }

    public int getAddress() {
        return address;
    }

    public void setAddress(int address) {
        this.address = address;
    }

    /*
     * Notifica a unidade
     */
    @Override
    public void updateCommunication(Object obj){

        try {


            if(obj instanceof  AnsCmd) {

                // Executa evento de recebmento de resposta de comando
                onReceiveAnswer((AnsCmd)obj);

            }else if(obj instanceof ChatMessage){

                // Evento recebimento menssagem do chat
                onReceiveChat((ChatMessage) obj);
            }
        }
        catch (Exception e) {

            System.out.println("Erro na execucao da callback pela unidade");
            System.out.println(e.toString());
            Log.d("",e.toString());
        }
    }

    @Override
    public void connEstablished(){

        communic = Communication.getInstance();
    }

    abstract public void onReceiveAnswer(AnsCmd ans);

    abstract public void onReceiveChat(ChatMessage chatMessage);
}
